src/main/java/com/eleks/academy/whoami/core/impl/TurnImpl.java

```diff
@@ -1,36 +1,94 @@
 package com.eleks.academy.whoami.core.impl;
 
-import java.util.List;
-
-import com.eleks.academy.whoami.core.Player;
+import com.eleks.academy.whoami.core.SynchronousPlayer;
 import com.eleks.academy.whoami.core.Turn;
+import com.eleks.academy.whoami.core.action.PlayerAction;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import java.util.Random;
 
 public class TurnImpl implements Turn {
-	
-	private final List<Player> players;
-	private int currentPlayerIndex = 0;
-	
-	public TurnImpl(List<Player> players) {
-		this.players = players;
+
+	private List<List<PlayerAction>> turns = new ArrayList<>();
+
+	public TurnImpl(List<SynchronousPlayer> players) {
+		SynchronousPlayer currentPlayer = players.get(new Random().nextInt(players.size()));
+		List<PlayerAction> playerActions = new ArrayList<>(players.size());
+
+		players.forEach(p -> {
+			if (currentPlayer.equals(p)) {
+				playerActions.add(new PlayerAction(p.getName(), PlayerAction.Action.QUESTION, null));
+			} else {
+				playerActions.add(new PlayerAction(p.getName(), PlayerAction.Action.ANSWER, null));
+			}
+		});
+		turns.add(playerActions);
+	}
+
+	@Override
+	public void makeTurn(List<SynchronousPlayer> players, boolean samePlayer) {
+		List<PlayerAction> playerActions = new ArrayList<>(players.size());
+		List<PlayerAction> previousTurn = turns.get(turns.size() - 1);
+
+		if (previousTurn.size() == players.size()) {
+			if (samePlayer) {
+				saveNewTurn(playerActions, previousTurn);
+			} else {
+				Collections.rotate(previousTurn, 1);
+
+				saveNewTurn(playerActions, previousTurn);
+			}
+		} else {
+			updateAvailablePlayers(previousTurn, players);
+
+			Collections.rotate(previousTurn, 1);
+
+			saveNewTurn(playerActions, previousTurn);
+		}
+		turns.add(playerActions);
 	}
-	
+
 	@Override
-	public Player getGuesser() {
-		return this.players.get(currentPlayerIndex);
+	public List<PlayerAction> getCurrentTurn() {
+		return this.turns.get(turns.size() - 1);
 	}
 
 	@Override
-	public List<Player> getOtherPlayers() {
-		return this.players.stream()
-				.filter(player -> !player.getName().equals(this.getGuesser().getName()))
-				.toList();
+	public List<List<PlayerAction>> getTurns() {
+		return this.turns; //TODO: history
 	}
-	
+
 	@Override
-	public void changeTurn() {
-		this.currentPlayerIndex = this.currentPlayerIndex + 1 >= this.players.size() ? 0 : this.currentPlayerIndex + 1; 
+	public void action(String player, PlayerAction action, String question) {
```

Comment:
please implement this or or throw exception not supported method, never leave method empty